File: scripts/catalog_query.py
#!/usr/bin/env python3
"""What the catalog says, without the 1.12 MB CSV leaving the sandbox.

With --feed-ids it returns those rows. Without, it returns a summary, because 256 rows of
catalog metadata in a tool reply is the payload problem Code Mode exists to avoid.
Contact addresses are never in the output, only whether one is on file.
"""

import argparse
import json
from collections import Counter

from probe_catalog import load_catalog


def row_view(r):
    return {
        "feed_id": r.get("mdb_source_id"),
        "provider": r.get("provider"),
        "url": r.get("urls.direct_download"),
        "auth_type": r.get("urls.authentication_type") or "",
        "catalog_status": r.get("status") or "",
        "redirect_id": r.get("redirect.id") or "",
        "contact_on_file": bool(r.get("feed_contact_email")),
    }


def summary(rows):
    hosts = Counter(r["urls.direct_download"].split("/")[2].lower() for r in rows)
    return {
        "feeds": len(rows),
        "by_status": dict(Counter(r.get("status") or "unset" for r in rows)),
        "by_auth_type": dict(Counter(r.get("urls.authentication_type") or "unset" for r in rows)),
        "with_redirect": sum(1 for r in rows if r.get("redirect.id")),
        "with_contact_on_file": sum(1 for r in rows if r.get("feed_contact_email")),
        "top_hosts": hosts.most_common(10),
    }


def main():
    p = argparse.ArgumentParser()
    p.add_argument("--jurisdiction", default="California")
    p.add_argument("--feed-ids", help="comma-separated mdb_source_id list")
    a = p.parse_args()

    ids = a.feed_ids.split(",") if a.feed_ids else None
    rows = load_catalog(a.jurisdiction, ids)
    out = {"jurisdiction": a.jurisdiction, "summary": summary(rows)}
    if ids:
        out["rows"] = [row_view(r) for r in rows]
    print(json.dumps(out, indent=1))


if __name__ == "__main__":
    main()

File: packages/mcp/src/features/catalog/catalogQuery.ts
// catalog.load. The CSV is read in the sandbox and never crosses into a tool reply whole:
// without feed ids this is a summary, with them it is the rows asked for.
import { execFile } from 'node:child_process'
import { promisify } from 'node:util'

import { z } from 'zod'

import { env } from '../../utils/env.js'
import { fromRoot } from '../../utils/repoRoot.js'

const run = promisify(execFile)
const CATALOG_TIMEOUT_MS = 90_000
const MAX_OUTPUT_BYTES = 4 * 1024 * 1024
const SCRIPT = fromRoot('scripts/catalog_query.py')

export const catalogRowSchema = z.object({
  feed_id: z.string().nullable(),
  provider: z.string(),
  url: z.string(),
  auth_type: z.string(),
  catalog_status: z.string(),
  redirect_id: z.string(),
  contact_on_file: z.boolean(),
})

export const catalogResultSchema = z.object({
  jurisdiction: z.string(),
  summary: z.object({
    feeds: z.number().int(),
    by_status: z.record(z.string(), z.number()),
    by_auth_type: z.record(z.string(), z.number()),
    with_redirect: z.number().int(),
    with_contact_on_file: z.number().int(),
    top_hosts: z.array(z.tuple([z.string(), z.number()])),
  }),
  rows: z.array(catalogRowSchema).optional(),
})

export type CatalogResult = z.infer<typeof catalogResultSchema>

export async function queryCatalog(
  jurisdiction: string,
  feedIds?: readonly string[],
): Promise<CatalogResult> {
  const args = [SCRIPT, '--jurisdiction', jurisdiction]
  if (feedIds && feedIds.length > 0) args.push('--feed-ids', feedIds.join(','))
  const { stdout } = await run(env.CASEWORK_PYTHON, args, {
    timeout: CATALOG_TIMEOUT_MS,
    maxBuffer: MAX_OUTPUT_BYTES,
  })
  return catalogResultSchema.parse(JSON.parse(stdout))
}
